src/nacl/public.py

```diff
@@ -247,3 +247,86 @@ def shared_key(self):
+    a specified key-pair by using ephemeral sender's keypairs,
+    whose private part will be discarded just after encrypting
+    a single plaintext message.
+
+    The ciphertexts generated by :class:`~nacl.public.SecretBox` include
+    the public part of the ephemeral key before the :class:`~nacl.public.Box`
+    ciphertext.
+
+    :param public_key: :class:`~nacl.public.PublicKey` used to encrypt
+        messages and derive nonces
+    :param private_key: :class:`~nacl.public.PrivateKey` used to decrypt
+        messages
+    """
+
+    def __init__(self, recipient_key):
+
+        if isinstance(recipient_key, PublicKey):
+            self._public_key = recipient_key.encode(
+                                                encoder=encoding.RawEncoder)
+            self._private_key = None
+        elif isinstance(recipient_key, PrivateKey):
+            self._private_key = recipient_key.encode(
+                                                encoder=encoding.RawEncoder)
+            self._public_key = recipient_key.public_key.encode(
+                                                encoder=encoding.RawEncoder)
+        else:
+            raise exc.TypeError("SealedBox must be created from "
+                                "a PublicKey or a PrivateKey")
+
+    def __bytes__(self):
+        return self._public_key
+
+    def encrypt(self, plaintext, encoder=encoding.RawEncoder):
+        """
+        Encrypts the plaintext message using a random-generated ephemeral
+        keypair and returns the ciphertext encoded with the encoder.
+
+        The public part of the ephemeral key-pair will be enclosed in the
+        returned ciphertext.
+
+        The private part of the ephemeral key-pair will be scrubbed before
+        returning the ciphertext, therefore, the sender will not be able to
+        decrypt the generated ciphertext.
+
+        :param plaintext: [:class:`bytes`] The plaintext message to encrypt
+        :param encoder: The encoder to use to encode the ciphertext
+        :return bytes: encoded ciphertext
+        """
+
+        ciphertext = nacl.bindings.crypto_box_seal(
+            plaintext,
+            self._public_key
+        )
+
+        encoded_ciphertext = encoder.encode(ciphertext)
+
+        return encoded_ciphertext
+
+    def decrypt(self, ciphertext, encoder=encoding.RawEncoder):
+        """
+        Decrypts the ciphertext using the ephemeral public key enclosed
+        in the ciphertext and the SealedBox private key, returning
+        the plaintext message.
+
+        :param ciphertext: [:class:`bytes`] The encrypted message to decrypt
+        :param encoder: The encoder used to decode the ciphertext.
+        :return bytes: The original plaintext
+        """
+        # Decode our ciphertext
+        ciphertext = encoder.decode(ciphertext)
+
+        plaintext = nacl.bindings.crypto_box_seal_open(
+            ciphertext,
+            self._public_key,
+            self._private_key,
```

Comment:
This will raise a cffi exception if the user instantiates `SealedBox` with a public key and then attempts to decrypt. We probably should raise a TypeError if you call decrypt in that case.